### PR TITLE
Add a litmusbook for cruncy-postgres workload

### PR DESCRIPTION
**Pull request by shashank855**
**This PR does the following**:
- Adds a litmusbook which trigers a crunchy-postgres-loadgen job.
- The litmusbook checks for the following after deploying the loadgen:
  1. Weather the loadgen pod is in running state.
  2. Weather the loadgen has perfomed a write operation on database or not.
- Accordingly ansible task updates the result in result CR.
 
**Special note for Reviewers**
- The state-ful set for the mongodb deployment is also modified to make the mongo workload work properly
  refer:https://github.com/openebs/test-tools/pull/12

**Following is the output acheived after deploying the loadgen:**
```
PLAY [localhost] ***************************************************************

TASK [Generate the litmus result CR to reflect SOT (Start of Test)] ************

TASK [Apply the litmus result CR] **********************************************

TASK [Checking the status  of test specific namespace.] ************************

TASK [Create Cassandra Loadgen Job] ********************************************
FAILED - RETRYING: Verify loadgen pod is running (15 retries left).

TASK [Verify loadgen pod is running] *******************************************

TASK [Verifying Load-generation] ***********************************************

TASK [Generate the litmus result CR to reflect EOT (End of Test)] **************

TASK [Apply the litmus result CR] **********************************************

PLAY RECAP *********************************************************************
127.0.0.1                  : ok=10   changed=8    unreachable=0    failed=0   

```
**Following is the output of load-generation:**
```
spawn pgbench -i -h pgset.litmus.svc.cluster.local -p 5432 -U testuser -s 30 postgres
Password: 
password

NOTICE:  table "pgbench_history" does not exist, skipping
NOTICE:  table "pgbench_tellers" does not exist, skipping
NOTICE:  table "pgbench_accounts" does not exist, skipping
NOTICE:  table "pgbench_branches" does not exist, skipping
creating tables...
100000 of 3000000 tuples (3%) done (elapsed 0.05 s, remaining 1.44 s)
200000 of 3000000 tuples (6%) done (elapsed 0.51 s, remaining 7.11 s)
300000 of 3000000 tuples (10%) done (elapsed 0.95 s, remaining 8.54 s)
400000 of 3000000 tuples (13%) done (elapsed 1.04 s, remaining 6.75 s)
500000 of 3000000 tuples (16%) done (elapsed 1.44 s, remaining 7.21 s)
600000 of 3000000 tuples (20%) done (elapsed 1.55 s, remaining 6.18 s)
700000 of 3000000 tuples (23%) done (elapsed 1.86 s, remaining 6.10 s)
800000 of 3000000 tuples (26%) done (elapsed 2.25 s, remaining 6.19 s)
900000 of 3000000 tuples (30%) done (elapsed 2.34 s, remaining 5.47 s)
1000000 of 3000000 tuples (33%) done (elapsed 3.13 s, remaining 6.25 s)
1100000 of 3000000 tuples (36%) done (elapsed 3.32 s, remaining 5.74 s)
1200000 of 3000000 tuples (40%) done (elapsed 3.58 s, remaining 5.37 s)
1300000 of 3000000 tuples (43%) done (elapsed 4.34 s, remaining 5.67 s)
1400000 of 3000000 tuples (46%) done (elapsed 4.57 s, remaining 5.22 s)
1500000 of 3000000 tuples (50%) done (elapsed 4.85 s, remaining 4.85 s)
1600000 of 3000000 tuples (53%) done (elapsed 5.95 s, remaining 5.21 s)
1700000 of 3000000 tuples (56%) done (elapsed 6.13 s, remaining 4.69 s)
1800000 of 3000000 tuples (60%) done (elapsed 6.40 s, remaining 4.27 s)
1900000 of 3000000 tuples (63%) done (elapsed 6.75 s, remaining 3.91 s)
2000000 of 3000000 tuples (66%) done (elapsed 7.49 s, remaining 3.75 s)
2100000 of 3000000 tuples (70%) done (elapsed 8.24 s, remaining 3.53 s)
2200000 of 3000000 tuples (73%) done (elapsed 8.68 s, remaining 3.16 s)
2300000 of 3000000 tuples (76%) done (elapsed 8.89 s, remaining 2.71 s)
spawn pgbench -c 4 -h pgset.litmus.svc.cluster.local -p 5432 -U testuser -j 5 -t 200 postgres
Password: 

starting vacuum...end.
spawn pgbench -c 4 -h pgset.litmus.svc.cluster.local -p 5432 -U testuser -j 5 -t 200 postgres
Password: 
starting vacuum...end.
spawn pgbench -c 4 -h pgset.litmus.svc.cluster.local -p 5432 -U testuser -j 5 -t 200 postgres
Password: 

starting vacuum...end.
transaction type: <builtin: TPC-B (sort of)>
scaling factor: 30
query mode: simple
number of clients: 4
number of threads: 4
number of transactions per client: 200
number of transactions actually processed: 800/800
latency average = 24.736 ms
tps = 161.704601 (including connections establishing)
tps = 162.821928 (excluding connections establishing)
spawn pgbench -c 4 -h pgset.litmus.svc.cluster.local -p 5432 -U testuser -j 5 -t 200 postgres
Password: 
starting vacuum...end.
transaction type: <builtin: TPC-B (sort of)>
scaling factor: 30
query mode: simple
number of clients: 4
number of threads: 4
number of transactions per client: 200
number of transactions actually processed: 800/800
latency average = 5.522 ms
tps = 724.385802 (including connections establishing)
tps = 734.919177 (excluding connections establishing)
spawn pgbench -c 4 -h pgset.litmus.svc.cluster.local -p 5432 -U testuser -j 5 -t 200 postgres
Password: 
starting vacuum...end.
transaction type: <builtin: TPC-B (sort of)>
scaling factor: 30
query mode: simple
number of clients: 4
number of threads: 4
number of transactions per client: 200
number of transactions actually processed: 800/800
latency average = 6.196 ms
tps = 645.625025 (including connections establishing)
tps = 663.130595 (excluding connections establishing)
spawn pgbench -c 4 -h pgset.litmus.svc.cluster.local -p 5432 -U testuser -j 5 -t 200 postgres
Password: 
starting vacuum...end.
transaction type: <builtin: TPC-B (sort of)>
scaling factor: 30
query mode: simple
number of clients: 4
number of threads: 4
number of transactions per client: 200
number of transactions actually processed: 800/800
latency average = 5.488 ms
tps = 728.816841 (including connections establishing)
tps = 735.939821 (excluding connections establishing)
spawn pgbench -c 4 -h pgset.litmus.svc.cluster.local -p 5432 -U testuser -j 5 -t 200 postgres
Password: 
starting vacuum...end.
transaction type: <builtin: TPC-B (sort of)>
scaling factor: 30
query mode: simple
number of clients: 4
number of threads: 4
number of transactions per client: 200
number of transactions actually processed: 800/800
latency average = 5.798 ms
tps = 689.949885 (including connections establishing)
tps = 711.724228 (excluding connections establishing)
```

### CODE REVIEW

**Comment by shashank855** [comment]
@dargasudarshan  I have resolved the conflict.

File: apps/crunchy-postgres/workload/test.yml [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        - name: Checking the status  of test specific namespace.
+          shell: kubectl get ns {{ namespace }} -o jsonpath='{.status.phase}'
+          args:
+           executable: /bin/bash
+          register: npstatus
+          until: "'Active' in npstatus.stdout"
+          delay: 30
+          retries: 10
+        
+        - name: Create Cassandra Loadgen Job
```
Comment (dargasudarshan) [inline apps/crunchy-postgres/workload/test.yml]: Rename the task name from Cassandra Loadgen to Crunchy Loadgen

Comment (shashank855) [reply apps/crunchy-postgres/workload/test.yml]: @dargasudarshan  sure

File: apps/crunchy-postgres/workload/test.yml [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+        - name: Verifying Load-generation
+          shell: kubectl exec -it pgset-0 -n {{ namespace }} -- psql -c '\dt'
+          register: output
+          until: "'pgbench_accounts' in output.stdout"
+          delay: 30
+          retries: 15
+
+        - set_fact:
+            flag: "loadgen_started"
```
[the remark this answers is not in the document]
Comment (shashank855) [reply apps/crunchy-postgres/workload/test.yml]: @ksatchit Sure

**Comment by shashank855** [comment]
@ksatchit I have made the required changes

File: apps/mongodb/deployers/mongo_statefulset.yml [hunk 3]
```diff
@@ -31,12 +31,12 @@ spec:
...
-           - rs0
-           - "--smallfiles"
-           - "--noprealloc"
-           - "--bind_ip_all"
+           #- mongod
+           #- "--replSet"
+           #- rs0
+           #- "--smallfiles"
+           #- "--noprealloc"
+           #- "--bind_ip_all"
```
Comment (gprasath) [inline apps/mongodb/deployers/mongo_statefulset.yml]: Why these lines are commented?

File: apps/crunchy-postgres/workload/run_litmus_test.yml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+      restartPolicy: Never
+      containers:
+      - name: loadgen
+        image: openebs/ansible-runner
+        imagePullPolicy: Always
+        env: 
+        
+          - name: ANSIBLE_STDOUT_CALLBACK
+            #value: log_plays
+            value: actionable
```
Comment (gprasath) [inline apps/crunchy-postgres/workload/run_litmus_test.yml]: use default

File: apps/crunchy-postgres/workload/crunchy_postgres_loadgen.yml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+          - name: PORT
+            value: "5432"
+            
+            #Number of parallel transactions to perform
+          - name: PARALLEL_TRANSACTION
+            value: "5"
+            
+            #Number of transaction to perform
+          - name: TRANSACTIONS
+            value: "200"
```
Comment (gprasath) [inline apps/crunchy-postgres/workload/crunchy_postgres_loadgen.yml]: pass all the above envs from litmusbook

File: apps/crunchy-postgres/workload/test.yml [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+          register: npstatus
+          until: "'Active' in npstatus.stdout"
+          delay: 30
+          retries: 10
+        
+        - name: Create Crunchy-postgres Loadgen Job
+          shell: kubectl apply -f {{ crunchy_loadgen }} -n {{ namespace }} 
+
+        - name: Verify loadgen pod is running 
+          shell: kubectl get pods -n litmus -l app=crunchy-loadgen -o jsonpath='{.items[0].status.phase}'
```
Comment (gprasath) [inline apps/crunchy-postgres/workload/test.yml]: Use {{ namesapce }}

File: apps/crunchy-postgres/workload/run_litmus_test.yml [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+metadata:
+  name: cruncy-postgres-loadgen 
+  namespace: litmus
+spec:
+  template:
+    metadata:
+      name: crunchy_postgres-loadgen
+      namespace: litmus
+      labels:
+        app: crunchy-postgres-loadgen
```
Comment (gprasath) [inline apps/crunchy-postgres/workload/run_litmus_test.yml]: Use the label key as 'loadgen' for load generator pod in all the load generators.

File: apps/mongodb/deployers/mongo_statefulset.yml (hunk 3 above)
Comment (shashank855) [reply apps/mongodb/deployers/mongo_statefulset.yml]: @gprasath We don't need these commands. We are letting mongodb to use the default configuration for sever initiation, so to perform our loadgen operation correctly.

File: apps/crunchy-postgres/workload/run_litmus_test.yml (hunk 7 above)
Comment (shashank855) [reply apps/crunchy-postgres/workload/run_litmus_test.yml]: @gprasath We cant use same app labels for all loadgens..there may be a scenario where two loadgens running on same cluster...then in that case there will be a contradiction.

Comment (shashank855) [reply apps/crunchy-postgres/workload/run_litmus_test.yml]: @gprasath I have made this change in actual loadgen job

**Comment by shashank855** [comment]
#215

**Pull request closed without merging**